libraries/Parsers/Microsoft.Bot.Builder.Parsers.LU/Parser/EntitySection.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+using System.Collections.Generic;
+
+namespace Microsoft.Bot.Builder.Parsers.LU.Parser
+{
+    /// <summary>SectionEntity class.</summary>
+    public partial class SectionEntity
+    {
+        // TODO: pass this constant to a helper class.
+        private readonly char[] _invalidCharsInIntentOrEntityName = { '<', '>', '*', '%', '&', ':', '\\', '$' };
+
+        public SectionEntity(LUFileParser.EntitySectionContext parseTree)
+        {
+            SectionType = SectionType.EntitySection;
+            Errors = new List<Error>();
+            Name = ExtractName(parseTree);
+            Type = ExtractType(parseTree);
+            SynonymsOrPhraseList = ExtractSynonymsOrPhraseList(parseTree);
+            string secTypeStr = $"{SectionType}";
+            Id = $"{char.ToLowerInvariant(secTypeStr[0]) + secTypeStr.Substring(1)}_{Name}";
+            var startPosition = new Position { Line = parseTree.Start.Line, Character = parseTree.Start.Column };
+            var stopPosition = new Position { Line = parseTree.Stop.Line, Character = parseTree.Stop.Column + parseTree.Stop.Text.Length };
+            Range = new Range { Start = startPosition, End = stopPosition };
+        }
+
+        public string ExtractName(LUFileParser.EntitySectionContext parseTree)
+        {
+            var entityName = string.Empty;
+            if (parseTree.entityDefinition().entityLine().entityName() != null)
+            {
+                entityName = parseTree.entityDefinition().entityLine().entityName().GetText().Trim();
+            }
+            else
+            {
+                Errors.Add(
+                    Diagnostic.BuildDiagnostic(
+                        message: "Invalid entity line, did you miss entity name after $?",
+                        context: parseTree.entityDefinition().entityLine()));
+            }
+
+            if (!string.IsNullOrEmpty(entityName) && entityName.IndexOfAny(_invalidCharsInIntentOrEntityName) >= 0)
+            {
+                Errors.Add(
+                    Diagnostic.BuildDiagnostic(
+                        message: $"Invalid entity line, entity name {entityName} cannot contain any of the following characters: [<, >, *, %, &, :, \\, $]",
+                        context: parseTree.entityDefinition().entityLine()));
+                return null;
+            }
+            else
+            {
+                return entityName;
+            }
+        }
+
+        public string ExtractType(LUFileParser.EntitySectionContext parseTree)
+        {
+            if (parseTree.entityDefinition().entityLine().entityType() != null)
```

Comment:
Comment here. what is the high level logic for this if / else? This area needs more code for future people to have a semantic undersanding of what is going on.